Add v2.0 workflows API

brianparry opened the pull request:
* GET, POST, DELETE, cancel a workflow.
* GET, PUT, DELETE a graph definition.
* GET, PUT, DELETE a task definition.
* active query param to workflows route

Separate the workflows routes into workflows /workflows/graphs and /workflows/tasks.  The graphs route is used for graph definitions. ant the tasks route is used for task definitions  Some semantics are now inconsistent with /nodes/:identifier/workflows.  This will be cleaned up in another PR.

See https://groups.google.com/forum/#!topic/rackhd/SvggYYKeeaQ for more detailed discussion.

This change has contributions from @dalebremner @abildw and @srinia6 

Depends on https://github.com/RackHD/on-core/pull/107

abildw commented:
+1

brianparry commented:
test this please

brianparry commented:
test this please

brianparry commented:
test this please

benbp commented:
Resolves https://github.com/RackHD/RackHD/issues/95

benbp commented on lib/services/workflow-api-service.js:
```diff
@@ -155,7 +237,32 @@ function workflowApiServiceFactory(
+        return waterline.graphobjects.find({ node: target })
+            .then(function(workflows){
+                return _.filter(workflows, function(wf) {
+                    return wf.active();
```
Why do this work in memory rather than letting the query in `findActiveGraphForTarget` tell the database to do all the processing?

brianparry replied:
@dalebremner Correct me if I'm wrong, but I think the idea was to make this a bit more behavioral.  The code reads as "filter for active workflows" rather than "query for _status equals pending".  
The reason for not using the task graph store is that `taskGraphStore.findActiveGraphForTarget` goes directly to Mongo which results in returned resources that are inconsistent with the /workflows and /workflows/:id

benbp replied:
In this case though, `_status = "pending" and node = target` is the underlying criteria for an active workflow for the node, and it will only ever be zero or one objects, whereas doing a find all has no upper bound and could potentially be very slow. I get that it's easier to understand what's happening from the `return wf.active()` statement, but that was also my intent with the method name of `findActiveGraphForTarget`.

Re your second point, is there perhaps a way to cast the objects coming direct out of mongo to resemble the objects out of the waterline layer (I thought there wasn't a difference, but if there is we should probably be doing this by default)? What are the differences in the resources?

brianparry replied:
@benbp Yeah, I agree a query makes sense here.  I think the thing that makes this a little confusing is that _status == pending is synonomous with "active".  Maybe we should add an active property to Constants.Task.States?

On the second point, I think the difference was the id property.  The raw Mongo document named this property "_id" whereas waterline omits the underscore.  There may be other differences too, but I'd need to look at it again.

benbp replied:
@brianparry I agree, I think the right thing to do might be to change the state to `Constants.Task.States.Running` since it's a pre-existing state? We didn't do it originally because it saved a query operation, but I think it makes sense given how confusing it is.

Ah ok, yeah I remember making some mistakes/inconsistencies with `id` vs. `_id`, so it's possible I didn't resolve all of them.

benbp commented:
:+1:

brianparry commented:
https://github.com/RackHD/on-core/pull/107 is merged, so test this please.

brianparry commented:
Once again, test this please

yyscamper commented on lib/api/2.0/workflowGraphs.js:
```diff
@@ -0,0 +1,67 @@
+var injector = require('../../../index.js').injector;
+var controller = injector.get('Http.Services.Swagger').controller;
+var workflowApiService = injector.get('Http.Services.Api.Workflows');
+var _ = require('lodash');    // jshint ignore:line
```
Since `lodash` has already in the injectable tables, so how about using the `injector.get` instead of require it directly?

I am thinking about a possible case to reduce the RackHD footprint in future. If the lodash have already in the on-core's injectable tables, so other modules (on-http, on-taskgraph, on-tasks, ...) can always obtain it via on-core, thus the `lodash` can be removed from on-http's dependencies, avoid to install a duplicated lodash in on-http. 

My point is that if the module has been included in on-core, then we should leverage the on-core to load it.

yyscamper commented on lib/api/2.0/workflows.js:
```diff
@@ -0,0 +1,112 @@
+
+    throw new Errors.BadRequestError(
+        command + ' is not a valid workflow action'
+    );
```
If only talk for the functionality of this function, I think your code is right and works well. I just have a little concern for this function's logic because it may give user the first impression that this function will eventually throw an error.
My point is usually the error handling is putting into the branching, not in the main logic flow.
```javascript
if (!_(actionFunctions).has(command)) {
    throw new Errors.BadRequestError(...);
}
return actionFunctions[command]();
```
or put the error handling in the `else`:
```javascript
if (_(actionFunctions).has(command)) {
    return actionFunctions[command]();
} else {
    throw new Errors.BadRequestError(...);
}
```

This may be my personal style, so this is not a blocking problem for me.  But I want to hear your comment about this.

brianparry replied:
@yyscamper Yep - usually I structure code like this like your first example.  I prefer this to including error handling in a conditional because I think it obscures your mainline code.  In this case, I structured it more like a default case in a switch statement.  I'll change it to look like your first suggestion.

yyscamper commented on lib/services/workflow-api-service.js:
```diff
@@ -128,7 +128,56 @@ function workflowApiServiceFactory(
+            }
+            throw new Errors.TaskCancellationError(
+                graphId + ' is not an active workflow'
+            );
```
This is another case for my above concern about putting error handling into branch. But this case is a little nicer than above because it is not in the end of the function.

brianparry replied:
I'll change it to this to be consistent with the first case.
```javascript
          if (!workflow.active()) {
               throw new Errors.TaskCancellationError(
                    graphId + ' is not an active workflow'
               );
          }
          return taskGraphProtocol.cancelTaskGraph(instanceId);

yyscamper commented on lib/services/workflow-api-service.js:
```diff
@@ -128,7 +128,56 @@ function workflowApiServiceFactory(
+            return instanceId;
+        })
+        .then(function(cancelledGraphId) {
+            if (cancelledGraphId !== instanceId) {
```
I think this judgment is redundant, base on your design, the `cancelledGraphId` and `instanceId` is always the equivalent. So the next throwing will never be executed and this is always a coverage hole for unit-testing.

brianparry replied:
If the cancel fails, it will return null for cancelled graphId.

yyscamper replied:
ohh, yes, you are right, forget my this comment.

yyscamper commented on lib/services/workflow-api-service.js:
```diff
@@ -155,7 +236,22 @@ function workflowApiServiceFactory(
+        return waterline.graphobjects.needByIdentifier(id);
+    };
+
+    WorkflowApiService.prototype.destroyGraphDefinition = function(injectableName) {
```
How above to use the uniform words for the same action of task definition and graph definition?
This is your current mapping:
```
destroyGraphDefinition <--> deleteWorkflowsTasksByName 
getGraphDefinitions  <--> getWorkflowsTasksByName 
defineTaskGraph <--> putWorkflowsTasksByName
```
If we use `destroy` for graph, then it should be better to use `destroy` for task as well; If we add `byName` for task, then it should also be better to add the words for graph.

brianparry replied:
I agree with this.  As above, I think we should consider this a separate task that we track with a Github issue.

yyscamper commented on lib/services/workflow-api-service.js:
```diff
@@ -155,7 +236,22 @@ function workflowApiServiceFactory(
+        });
+    };
+
+    WorkflowApiService.prototype.getActiveWorkflows = function(query) {
```
This API name is not well-matched with its implementation. If the API automatically builds the query to lookup active workflow, then I think the name and implementation is matched well.

brianparry replied:
I'm not sure I follow your concern.  Are you suggesting that _graph_ be replaced with _workflow_?  If so, I agree and I think this applies to many of the functions in this service.  I'd suggest adding an issue and cleaning up in the future.

yyscamper replied:
my meaning is that your function is more like "find workflow by query" not "find the active workflow", if user doesn't input the correct query, then your function works normally but not return the active one, this doesn't match with the function name.

brianparry replied:
I see what you mean now ().  The problem is that this function returns all workflows that match the provided query, not just active workflows.  Again, I think most of these functions need to be renamed to follow a consistent convention.

yyscamper commented on spec/lib/api/2.0/workflowGraphs-spec.js:
```diff
@@ -0,0 +1,89 @@
+
+            return helper.request().put('/api/2.0/workflows/graphs')
+            .send(graph)
+            .expect('Content-Type', /^application\/json/)
```
Also a consistent style concern, for some test cases, this `expect` has 4-spaces indent but sometimes not.

brianparry replied:
I think this is because several developers authored this code.  I think the style varies by developer preference.

benbp replied:
Yeah, I believe I may be the culprit for starting the no indent fad >;)

brianparry replied:
Haha @benbp!  I could go either way on this.  I think it looks better _with_ an indent if there are no callback functions passed to the chained methods.
```javascript
expect(foo).to.have.property(bar)
    .and.to.equal(baz)
```

I think it looks better _without_ the indent if there are callback functions.

```javascript
    var foo = _(bar)
    .filter(function(baz) {
        return baz === qux;
    });
```
  I'm consistently inconsistent!

yyscamper replied:
@brianparry @benbp do you think we should define a consistent style for this?so that we write code in same way.

brianparry replied:
@yyscamper I don't think there should be any requirement for us all to write code in the same style.  As long as the style is clear and expressive (such as the examples above) I'm fine with it.

yyscamper commented on spec/lib/services/workflow-api-service-spec.js:
```diff
@@ -250,4 +279,85 @@ describe('Http.Services.Api.Workflows', function () {
+                               id      : 'testgraphid',
+                               _status :"pending"
+
+                           };
```
A little strange alignment style.

yyscamper commented on spec/lib/services/workflow-api-service-spec.js:
```diff
@@ -250,4 +279,85 @@ describe('Http.Services.Api.Workflows', function () {
+    });
+
+
+
```
remove these unwanted lines.

yyscamper commented on spec/lib/services/workflow-api-service-spec.js:
```diff
@@ -250,4 +279,85 @@ describe('Http.Services.Api.Workflows', function () {
+    it('should return Not Found Error when invalid id is passed', function() {
+        waterline.graphobjects.needByIdentifier.rejects(new Errors.NotFoundError('Not Found'));
+        return expect(workflowApiService.getWorkflowById())
+               .to.be.rejectedWith(Errors.NotFoundError);
```
forgot 'eventually` for promise

yyscamper commented on spec/lib/services/workflow-api-service-spec.js:
```diff
@@ -250,4 +279,85 @@ describe('Http.Services.Api.Workflows', function () {
+        waterline.graphobjects.find.resolves(activeWorkflow);
+        return workflowApiService.getWorkflowById().then(function (workflows) {
+            expect(workflows).to.deep.equal(activeWorkflow);
+        });
```
Your writing is good, and this can be short to:
```javascript
return workflowApiService.getWorklfowById().to.become(activeWorkflow);
```

yyscamper commented on spec/lib/services/workflow-api-service-spec.js:
```diff
@@ -250,4 +279,85 @@ describe('Http.Services.Api.Workflows', function () {
+        );
+        waterline.graphobjects.needByIdentifier.rejects(mockWorkflowError);
+        return workflowApiService.cancelTaskGraph()
+            .should.eventually.be.rejectedWith(mockWorkflowError);
```
The `eventually` should be not written together with reject, please see this issue: https://github.com/RackHD/RackHD/issues/54

brianparry replied:
Will this work?
```javascript
return workflowApiService.cancelTaskGraph()
          .should.be.rejectedWith(mockWorkflowError);
```

or do I need to do this:

```javascript
return workflowApiService.cancelTaskGraph().catch(function(err) {
          expect(err).to.deepEqual(mockWorkflowError);
});
```

benbp replied:
Yeah the former should work fine I believe. The second doesn't test the case where an error isn't thrown, so the first block is the better options.

brianparry replied:
Changing to this

```javascript
return workflowApiService.cancelTaskGraph()
          .should.be.rejectedWith(mockWorkflowError);
```

yyscamper commented on spec/lib/api/2.0/workflowTasks-spec.js:
```diff
@@ -0,0 +1,173 @@
+                expect(res.body).to.have.property('injectableName', 'dummyName');
+                expect(res.body).to.have.property('options').to.be.an('object');
+                expect(res.body).to.have.deep.property('options.oids', 'SNMPv2-MIB::sysDescr');
+                });
```
incorrect indent

yyscamper commented:
Some of my comment is about the style, this is not  gating for me, but I think at least following two comments need be fixed:
1. Rename the name of workflow service APIs.
2. `eventually` should not be written together with rejected.

brianparry commented:
@yyscamper I agree that the function renaming should be done, but disagree that it should gate the PR.  I'd suggest logging an issue in github and cleaning up the naming later.  I addressed the `.eventually` issue and cleaned up the style in a few places.

yyscamper commented:
@brianparry , I'm ok to move the renaming to another PR, thanks for other change, so this PR is +1 for me.

yyscamper commented:
I just see the coverage reduce a lot, can you improve this?

yyscamper commented:
test this please

brianparry commented:
test this please

brianparry commented:
test this please
